feat(gateway): Add admin session authentication manager

f3r10 opened the pull request:
closes #4128 
# Motivation
Currently, authentication to the getaway admin endpoints is made through a password sent in clear text, which is a security concern.
This PR attempts to fix this issue by adding an Authentication Manager that distributes challenge codes that must be signed before users can request a JWT that allows them to authenticate to the gateway admin endpoints. 
Because of backward compatibility, the password check is not deleted but used as a fallback only if a JWT is absent in the request.
# This PR adds three additional endpoints:
- `/auth/challenge` generates a temporal challenge code
- `/auth/sign` once the challenge code is generated, this endpoint signs the challenge code
- `/auth/session` Once the challenge code is signed, it is possible to request a JWT token.
# How to test it:
- run `just mprocs`
- check which gateway address has to be used  running for example `gateway-lnd info`
- check which is the federation ID that has to be used with some commands such as `balance`
![image](https://github.com/user-attachments/assets/af451944-3857-4bd6-ab81-1b328d2aea23)
- Be aware that the process of requesting a challenge code, signing it, and requesting a JWT has to be done in less than one minute because of the expiration time.
- the JWT has an expiration time of 30 minutes
- `gateway-cli --address=http://127.0.0.1:13446/v1 get-challenge-auth`
- `gateway-cli --address=http://127.0.0.1:13446/v1 sign-challenge-auth --challenge-auth="02f0219bdc2ea9b4524a78f0a390863e16967940e9048171994b73d81305d76d7d-http://127.0.0.1:13446/v1-1728495691"`
- `gateway-cli --address=http://127.0.0.1:13446/v1 session-auth --challenge="02f0219bdc2ea9b4524a78f0a390863e16967940e9048171994b73d81305d76d7d-http://127.0.0.1:13446/v1-1728495691" --response="3acaa8ac174da093ff43eaab77e7ff916c202c8d744c2ec41df42f347b47ab03f162bb8c2a092008e804c90f9251d4553b95d8e2305d00e14012e19a25da6cb5"`
- `gateway-cli --address=http://127.0.0.1:13446/v1 --rpcjwt="eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZCI6ImlkLTY1MDQ5MDMxMDUzNzQxMzYzNTUiLCJleHAiOjE3Mjg0OTU2OTF9.mvBJcOLQqIZy5J8IDSr6SE9g9wSD0MiXbV9o6J9-30c" balance`

elsirion reviewed (commented):
Just some comments from a quick skim. Feels like work in progress, will mark it as draft.

Kodylow commented on gateway/ln-gateway/src/auth_manager.rs:
```diff
@@ -0,0 +1,183 @@
+    expiry: u64,
+}
+
+/// Structure to represent a challenge.
+/// gatewayid-gatewayapi-timestamp
+impl AuthChallenge {
+    /// Create a new challenge with a random string and an expiry time of 1
+    /// minute.
+    pub fn new(gateway_id: PublicKey, gateway_api: String) -> Self {
+        let now = fedimint_core::time::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("Time went backwards")
+            .as_secs();
+        let expiry = now + CHALLENGE_EXPIRY_SECONDS;
+        Self {
+            gateway_id,
+            gateway_api,
+            expiry,
+        }
+    }
+}
+
+impl FromStr for AuthChallenge {
+    type Err = anyhow::Error;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let parts = s.split('-').collect::<Vec<&str>>();
+        let gateway_id = PublicKey::from_str(parts[0])?;
+        let gateway_api = parts[1].to_string();
+        let expiry = parts[2].parse::<u64>()?;
+        Ok(Self {
+            gateway_id,
+            gateway_api,
+            expiry,
+        })
+    }
+}
```
this is for the challenge not for the actual jwt, it uses that crate there.

f3r10 commented:
> Just some comments from a quick skim. Feels like work in progress, will mark it as draft.

Thanks, @elsirion for the initial review, I addressed the comments and the PR will be ready for additional reviews.

otech47 reviewed (commented):
tested this and works as expected based on PR example steps

<img width="1110" alt="Screenshot 2024-11-08 at 11 41 44 AM" src="https://github.com/user-attachments/assets/d960ea63-d6c7-46d2-88fe-4039b27c7333">

I also tested that jwt session keys still work after generating a 2nd one. I generated 2 session keys and could use either of them to hit the balance

elsirion commented:
@elsirion review

Kodylow commented:
> tested this and works as expected based on PR example steps
> 
> <img alt="Screenshot 2024-11-08 at 11 41 44 AM" width="1110" src="https://private-user-images.githubusercontent.com/4914611/384409288-d960ea63-d6c7-46d2-88fe-4039b27c7333.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE5ODM0NzQsIm5iZiI6MTczMTk4MzE3NCwicGF0aCI6Ii80OTE0NjExLzM4NDQwOTI4OC1kOTYwZWE2My1kNmM3LTQ2ZDItODhmZS00MDM5YjI3YzczMzMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTExOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMTlUMDIyNjE0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Njc2NWE5NmE3OWIwZDg5MTM3NzI1MTNjNThlNjg5MTM2YWUzMGI4ZGZiN2Y4YTMzNzM3OTg3NGZiMTFhZjU4ZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.P5fOCLZNXCG3pLgnzv3ewgFdpy7n-hQQ_l7z4Seu-Lk">
> I also tested that jwt session keys still work after generating a 2nd one. I generated 2 session keys and could use either of them to hit the balance

the session keys just expire based on time, do we want to keep it to just a single one active ever?

m1sterc001guy commented:
> the session keys just expire based on time, do we want to keep it to just a single one active ever?

I think its fine to have two, maybe you want to access it from multiple devices. Can we rebase and run the CI?

f3r10 commented:
there are some errors on the `Build on linux` CI task that I am fixing now.

f3r10 commented:
closing this PR for a simple version #6535

f3r10 commented:
close this PR for #6535
